django/core/management/commands/shell.py

```diff
@@ -111,10 +115,33 @@ def python(self, options):
         # Start the interactive interpreter.
         code.interact(local=imported_objects)
 
+    def get_and_report_namespace(self, verbosity):
+        namespace = self.get_namespace()
+
+        if verbosity >= 1:
+            self.stdout.write(
+                f"{len(namespace)} objects imported automatically",
+                self.style.SUCCESS,
+            )
+
+    def get_namespace(self):
+        apps_models = apps.get_models()
+        apps_models_modules = [
+            (app.models_module, app.label) for app in apps.get_app_configs()
+        ]
+        namespace = {}
+        for model in reversed(apps_models):
+            if model.__module__:
+                namespace[model.__name__] = model
+        for app in apps_models_modules:
+            app_models_module, app_label = app
+            namespace[f"{app_label}_models"] = app_models_module
+        return namespace
+
     def handle(self, **options):
         # Execute the command and exit.
         if options["command"]:
-            exec(options["command"], globals())
+            exec(options["command"], {**globals(), **self.get_namespace()})
```

Comment:
Hmmm, thinking about it, I am not sure it makes sense to add `globals()` here, or below for stdin.

I inspected the contents by adding these lines at the top of `handle()`:

```
from pprint import pp
pp(globals())
```

The result:

```
$ ./manage.py shell
{'__name__': 'django.core.management.commands.shell',
 '__doc__': None,
 '__package__': 'django.core.management.commands',
 '__loader__': <_frozen_importlib_external.SourceFileLoader object at 0x101732c00>,
 '__spec__': ModuleSpec(name='django.core.management.commands.shell', loader=<_frozen_importlib_external.SourceFileLoader object at 0x101732c00>, origin='/Users/chainz/Documents/Projects/django/django/core/management/commands/shell.py'),
 '__file__': '/Users/chainz/Documents/Projects/django/django/core/management/commands/shell.py',
 '__cached__': '/Users/chainz/Documents/Projects/django/django/core/management/commands/__pycache__/shell.cpython-312.pyc',
 '__builtins__': {...},
 'os': <module 'os' (frozen)>,
 'select': <module 'select' (built-in)>,
 'sys': <module 'sys' (built-in)>,
 'traceback': <module 'traceback' from '/Users/chainz/.local/share/mise/installs/python/3.12.2/lib/python3.12/traceback.py'>,
 'BaseCommand': <class 'django.core.management.base.BaseCommand'>,
 'CommandError': <class 'django.core.management.base.CommandError'>,
 'OrderedSet': <class 'django.utils.datastructures.OrderedSet'>,
 'Command': <class 'django.core.management.commands.shell.Command'>}
```

I elided the `__builtins__` key, because it’s big, and `exec` [adds it by default](https://docs.python.org/3.12/library/functions.html#exec), if missing.

The remaining `__`/“dunder” names are related to the `shell` module, and so shouldn't really be used by passed-in code.

The non-dunder names (`os`, `sys`, etc.) are simply whatever `shell` happens to import. Naturally, they’ve changed over Django’s history.

If we drop the `**globals()` from these two places, all those names will go away, potentially breaking scripts using `-c` or stdin without explicit imports. But the namespace will be a lot more logical, and now would be the best time to change it. Users can always fix things by following the guide to override `get_namespace()` and add back the imports they rely on...

What do you think? Asking @DevilsAutumn too.

Reply:
Hmm, I think it would be better to remove `globals` to simplify things a bit in the namespace. If someone want any particular thing to import they can simply do it either externally or by overriding `get_namespace`.

One more thing that I still feel would be better to add is the [SHELL_IMPORTS](https://django-extensions.readthedocs.io/en/latest/shell_plus.html#additional-imports) settings. I see the the shell customization on 2 levels. 

At level 1, users would just want to import extra modules, submodules, methods etc. And I think this will be the most common use case.
At level 2, users would want to perform some kind of manipulation or do some pre-processing on their data. This one would be comparatively uncommon.

At this point, to achieve both levels, overriding `get_namespace` is the only option. But with `SHELL_IMPORTS`, we can super simplify things for atleast users at level 1(where most of the users lie?). As a developer, I would not want to create a bunch of files and folders, subclass a command and override a method just to import few methods/submodules/modules for testing. Rather I would prefer to write 3-4 lines in the settings.py and expect django to handle the complications of importing them.

Yes, If I had some complex operations to perform or want to use some other python runner before launching the shell, then I would definitely subclass the command and override a method.

Probably I should put this on the form as well to get opinions of a wider audience.

Reply:
At the end, have we decided to remove `globals`, or do we want to hear some feedback from the community?

Reply:
I think @DevilsAutumn and I are in agreement to remove the `globals()` calls here.

The `SHELL_IMPORTS` proposal is a second topic that should really be discussed in the forum. @salvo-polizzi would you like to start a post with a brief explanation of where the work is at, a description of the proposal to add a setting, and a quote from Bhuvnesh’s above comment?

I’m against adding the setting - happy to expand more on a forum thread.